src/engine.ml

```diff
@@ -1116,7 +1117,7 @@ let data_out ?add_timestamp (ctx : keys) hmac_algorithm compress protocol rng
     key data =
   (* as described in [out], ~add_timestamp is only used in static key mode *)
   let ctx, payload = out ?add_timestamp ctx hmac_algorithm compress rng data in
-  let out = Packet.encode protocol (key, `Data payload) in
+  let out = Packet.encode_data protocol key payload in
```

Comment:
this I guess is the path that we want to optimize the most.

right now, `outgoing` (below) is called with the data-to-be-sent. This calls `static_out` or `data_out`, which both call `out` followed by `encode_data` (or `encode_protocol`).

I'd like that `out` (which does a `Cstruct.concat`) already returns the (potentially) required prefix -- so in `encode_data` we won't need another allocation and blit. WDYT?

Reply:
Yes, sounds good! I'll take a look.

Reply:
I briefly looked into `out`, and came up with #218 -- may be worth to merge before modifying it further.